File: Linked_Lists/Medium/19/19.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        fast = head
        slow = head
        prev = None
        for i in range(n):
            if fast == None:
                return head
            fast = fast.next
        
        while fast != None:
            prev = slow
            slow = slow.next
            fast = fast.next

        if prev == None:
            head = head.next
        else:
            prev.next = slow.next

        return head

File: Linked_Lists/Medium/19/19.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
  let fast = head;
  let slow = head;
  let prev = null;

  for(let i = 0; i < n; i++){
    if(fast == null){
      return head;
    }
    fast = fast.next
  }

  while(fast){
    prev = slow
    slow = slow.next
    fast = fast.next
  }

  if(prev == null){
    head = head.next
  }else{
    prev.next = slow.next
  }
  return head
};